README
Ithumb-XS
=========

! ATTENTION !

The project was started for educational purposes (for learning Perl C API).
At this moment I haven't reasons for continued development.
There are more suitable solutions for creation of image thumbnails.
I close a project. Thanks all.

---

Ithumb::XS - is a small (one function), simple and fast Perl-XS module
for creation a thumbnails (with resizing and cropping), using Imlib2 library.

DEPENDENCES & REQUIRES

This module requires these other modules and libraries:

  Imlib2 (binary and headers)

Required Perl >= v5.24.0

INSTALLATION

MacOS prepare installation.

  brew cask install xquartz
  brew install imlib2

Manual installation.

To manual install this module type the following:

  perl Makefile.PL
  make
  make test
  make install

To automatical install this module type the following:

  cpan -i Ithumb::XS

USAGE

From shell (as one-liner):

  perl -MIthumb::XS=create_thumbnail -e \
    "create_thumbnail({width => 800, height => 600, src => '1.jpg', dst => '2.jpg'})"

From perl code:

  use Ithumb::XS 'create_thumbnail';

  create_thumbnail({
      width  => 800,
      height => 600,
      src    => 'src_image.jpg',
      dst    => 'dst_image.jpg',
  });

COPYRIGHT AND LICENCE

BSD 3-Clause License

Copyright (c) 2018-2021 Peter P. Neuromantic <p.brovchenko@protonmail.com>.
All rights reserved.

See LICENSE file for more information.

Changes
Revision history for Perl extension Ithumb::XS.

v0.5.5 2021-11-20
    - Removed executable flag on Makefile.PL, README*, etc
    - Fixed export function (create_image -> create_thumbnail)

v0.5.4 2021-11-19
    - Fixed CFLAGS override into Makefile.PL (thanks viy2 for bug report)
    - Added optimization option: -march=native
    - Removed flag: -Werror

v0.5.3 2020-07-04
    - Project is closed!
    - Fixed README*

v0.5.2 2020-03-01
    - Fixed README*
    - Fixed Cpants Kwalitee issues
    - Fixed XS.xs
      PROTOTYPES: DISABLE
    - Fixed Makefile.PL
      OPTIMIZE -O3
    - Fixed XS getting input values from hashref
      replace SvPV -> SvPV_nolen (suck less memory)

v0.5.1 2020-02-05
    - Fixed unused variables in ithumb.c
    - Fixed undeclared dependency Test::Pod::Coverage
      https://github.com/ChaoticEvil/p5-ithumb-xs/issues/2#issue-559945803
    - Added more strictly compiler flag (-Wall -Wextra -Werror -std=c11)

v0.5.0 2020-02-04
    - Added .travis.yml for CI over github
    - Added badges icons: build, coverage and cpan package version
    - Added cpanfile (for test pod coverage on 'make test')
    - Removed redundant OO-interface
    - Reverted method name convert_image -> create_thumbnail
    - Actualized README* and pod docs
    - Used wrap save-fucntion into more safely *with_error_return-fucntion
    - C and XS code changes:
      small refactoring;
      add func-wrapper for more safely saving result image;

v0.4.0 2019-12-04
    - Thumbnail algorithm moved to the library and separated from XS

v0.3.0 2019-11-29
    - Added new tests
    - Improved conversion algorithm
    - Renamed method (create_thumbnail -> convert_image)

v0.2.1 2019-02-17
    - Fixed pod-docs

v0.2.0 2019-02-17
    - Rewrote XS-code (params of thumbnail as hashref and better validation)
    - Added OO-interface: convert
    - Fixed tests
    - Fixed pod-doc ad readme

v0.1.5 2018-12-08
    - Fixed and improve readme's and pod-docs

v0.1.4.1 2018-11-22
    - Fixed render validation error
    - Fixed module description

v0.1.4 2018-11-22
    - Some XS improves (better input validation)

v0.1.3 2018-07-03
    - Improve XS code

v0.1.2 2018-03-14
    - Fixed CPAN Kwalitee Issues

v0.1.1 2018-03-11
    - Fixed test, pod-docs and typo.

v0.1.0 2018-03-11
    - The first release.
      Added basic functionality, create distribution and make cpan-package.


Makefile.PL
use 5.024000;
use ExtUtils::MakeMaker;
use Config;

WriteMakefile(
    NAME              => 'Ithumb::XS',
    ABSTRACT_FROM     => 'lib/Ithumb/XS.pm',
    VERSION_FROM      => 'lib/Ithumb/XS.pm',
    MIN_PERL_VERSION  => '5.24.0',
    PREREQ_PM         => {'XSLoader' => 0,},
    AUTHOR            => 'Peter P. Neuromantic <p.brovchenko@protonmail.com>',
    LICENSE           => 'bsd',
    LIBS              => ['-lImlib2'],
    DEFINE            => '',
    INC               => '-I/usr/include -I/usr/X11/include/',
    CCFLAGS           => ( $Config{ccflags} // '' ) . ' -Wall -Wextra -std=c11',
    OPTIMIZE          => '-O3 -march=native',
    META_MERGE        => {
        'meta-spec' => {
            version => 2
        },
        resources   => {
            bugtracker => {
                web => 'https://github.com/ChaoticEvil/p5-ithumb-xs/issues'
            },
            repository => {
                type => 'git',
                url  => 'https://github.com/ChaoticEvil/p5-ithumb-xs.git',
                web  => 'https://github.com/ChaoticEvil/p5-ithumb-xs',
            },
        },
    },
);


ithumb/ithumb.c
/*
 *
 * BSD 3-Clause License
 *
 * Copyright (c) 2018-2020 Peter P. Neuromantic <p.brovchenko@protonmail.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#include <stdio.h>
#include <string.h>
#include <Imlib2.h>

#include "ithumb.h"


IError get_error(int err_code) {
    IError error;

    switch(err_code) {
    case IMLIB_LOAD_ERROR_NONE:
        break;
    case IMLIB_LOAD_ERROR_FILE_DOES_NOT_EXIST:
        error.msg = "[Ithumb::XS] imlib error: File does not exist";
        break;
    case IMLIB_LOAD_ERROR_FILE_IS_DIRECTORY:
        error.msg = "[Ithumb::XS] imlib error: File is directory";
        break;
    case IMLIB_LOAD_ERROR_PERMISSION_DENIED_TO_READ:
        error.msg = "[Ithumb::XS] imlib error: Permission denied";
        break;
    case IMLIB_LOAD_ERROR_NO_LOADER_FOR_FILE_FORMAT:
        error.msg = "[Ithumb::XS] imlib error: No loader for file format";
        break;
    case IMLIB_LOAD_ERROR_PATH_TOO_LONG:
        error.msg = "[Ithumb::XS] imlib error: Path too long";
        break;
    case IMLIB_LOAD_ERROR_PATH_COMPONENT_NON_EXISTANT:
        error.msg = "[Ithumb::XS] imlib error: Path component non existant";
        break;
    case IMLIB_LOAD_ERROR_PATH_COMPONENT_NOT_DIRECTORY:
        error.msg = "[Ithumb::XS] imlib error: Path component not directory";
        break;
    case IMLIB_LOAD_ERROR_PATH_POINTS_OUTSIDE_ADDRESS_SPACE:
        error.msg = "[Ithumb::XS] imlib error: Path points outside address space";
        break;
    case IMLIB_LOAD_ERROR_TOO_MANY_SYMBOLIC_LINKS:
        error.msg = "[Ithumb::XS] imlib error: Too many symbolic links";
        break;
    case IMLIB_LOAD_ERROR_OUT_OF_MEMORY:
        error.msg = "[Ithumb::XS] imlib error: Out of memory";
        break;
    case IMLIB_LOAD_ERROR_OUT_OF_FILE_DESCRIPTORS:
        error.msg = "[Ithumb::XS] imlib error: Out of file descriptors";
        break;
    case IMLIB_LOAD_ERROR_PERMISSION_DENIED_TO_WRITE:
        error.msg = "[Ithumb::XS] imlib error: Permission denied to write";
        break;
    case IMLIB_LOAD_ERROR_OUT_OF_DISK_SPACE:
        error.msg = "[Ithumb::XS] imlib error: Out of disk space";
        break;
    case IMLIB_LOAD_ERROR_UNKNOWN:
        error.msg = "[Ithumb::XS] imlib error: Unknown";
        break;
    case ITHUMB_WIDTH_ERROR:
        error.msg = "[Ithumb::XS] error: invalid value of width (width must be a positive integer)";
        break;
    case ITHUMB_HEIGHT_ERROR:
        error.msg = "[Ithumb::XS] error: invalid value of height (height must be a positive integer)";
        break;
    case ITHUMB_SRC_ERROR:
        error.msg = "[Ithumb::XS] error: invalid value of source file path";
        break;
    case ITHUMB_DST_ERROR:
        error.msg = "[Ithumb::XS] error: invalid value of destination file path";
        break;
    case ITHUMB_SCALING_ERROR:
        error.msg = "[Ithumb::XS] error: image can't be a scaled";
        break;
    case ITHUMB_CROPPING_ERROR:
        error.msg = "[Ithumb::XS] error: image can't be croped";
        break;
    default:
        break;
    }

    error.code = err_code;

    return error;
}

int resize_and_crop(Image *image) {
    float aspect_ration_orig, aspect_ratio_new;

    int width = 0, height = 0;

    Imlib_Image src_img, scaled_img, croped_img;
    Imlib_Load_Error imlib_error = IMLIB_LOAD_ERROR_NONE;

    if (image->w <= 0)
        return ITHUMB_WIDTH_ERROR;

    if (image->h <= 0)
        return ITHUMB_HEIGHT_ERROR;

    if (!strlen(image->src))
        return ITHUMB_SRC_ERROR;

    if (!strlen(image->dst))
        return ITHUMB_DST_ERROR;

    src_img = imlib_load_image_with_error_return(image->src, &imlib_error);

    if (imlib_error)
        return imlib_error;

    imlib_context_set_image(src_img);

    width  = imlib_image_get_width();
    height = imlib_image_get_height();
    aspect_ration_orig = (float)width / (float)height;
    aspect_ratio_new = (float)image->w / (float)image->h;

    if ( aspect_ratio_new > 1 ) {
        scaled_img = imlib_create_cropped_scaled_image(
            0, 0, width, height, (int)(image->h * aspect_ration_orig), image->h);
    } else {
        scaled_img = imlib_create_cropped_scaled_image(
            0, 0, width, height, (int)(image->h * aspect_ration_orig), image->h);
    }

    scaled_img = imlib_create_cropped_scaled_image(
        0, 0, width, height, 640, 360);

    if (!scaled_img)
        return ITHUMB_SCALING_ERROR;

    imlib_context_set_image(scaled_img);

    croped_img = imlib_create_cropped_image(0, 0, image->w, 200);

    if (!croped_img)
        return ITHUMB_CROPPING_ERROR;

    imlib_context_set_image(croped_img);
    imlib_save_image_with_error_return(image->dst, &imlib_error);

    if (imlib_error)
        return imlib_error;

    return 0;
}


lib/Ithumb/XS.pm
package Ithumb::XS;

use 5.024000;
use strict;
use warnings;

use base 'Exporter';
our @EXPORT_OK = qw(create_thumbnail);

our $VERSION = '0.5.5';

require XSLoader;
XSLoader::load('Ithumb::XS', $VERSION);

1;

__END__

=encoding UTF-8

=head1 NAME

Ithumb::XS - Image thumbnail creation routines

=head1 DESCRIPTION

Ithumb::XS - is a small (one function), simple and fast Perl-XS module
for creation a thumbnails (with resizing and cropping), using Imlib2 library.

=head1 AUTHOR

Peter P. Neuromantic <p.brovchenko@protonmail.com>

=head1 SYNOPSIS

  use Ithumb::XS ();

  Ithumb::XS::create_thumbnail({
      width  => 800,
      height => 600,
      src    => 'src_image.jpg',
      dst    => 'dst_image.jpg'
  });

=head1 FUNCTIONS

=head2 create_thumbnail($)

Creates a small copy (resizing and cropping) of the image.

=over 20

=item $_[0]->{width}  - destination width

=item $_[0]->{height} - destination height

=item $_[0]->{src}    - path to the source image

=item $_[0]->{dst}    - path to the destionation result image

=back

=head1 LICENSE AND COPYRIGHT

BSD 3-Clause License

Copyright (c) 2018-2021 Peter P. Neuromantic E<lt>p.brovchenko@protonmail.comE<gt>.
All rights reserved.

See LICENSE file for more information.

=cut
